packages/@aws-cdk/aws-iam/lib/managed-policy.ts

```diff
@@ -252,6 +255,28 @@ export class ManagedPolicy extends Resource implements IManagedPolicy {
       resourceName: this.physicalName,
     });
 
+    const self = this;
+
+    class GrantPrincipal implements IPrincipal {
+      public readonly grantPrincipal: IPrincipal = this;
+
+      public get assumeRoleAction(): string {
```

Comment:
If your functions are only throwing errors, this is the wrong level of abstraction.